spec.md

```diff
@@ -2734,6 +2780,269 @@ message NodeExpandVolumeResponse {
+  // snapshot creation request. This field is needed in case the
+  // volume level secrets are different from the above secrets
+  // for the group snapshot.
+  // This field is OPTIONAL.
+  repeated VolumeSecret volume_secrets = 4;
+
+  // Plugin specific parameters passed in as opaque key-value pairs.
+  // This field is OPTIONAL. The Plugin is responsible for parsing and
+  // validating these parameters. COs will treat these as opaque.
+  map<string, string> parameters = 5;
+}
+
+message VolumeSecret {
+  // ID of the volume whose secrets are provided.
+  // This field is REQUIRED.
+  string volume_id = 1;
+
+  // Secrets required by plugin for a volume operation.
+  // This field is REQUIRED. Refer to the `Secrets Requirements`
+  // section on how to use this field.
+  map<string, string> secrets = 2 [(.csi.v1.csi_secret) = true];
+}
+
+message CreateVolumeGroupSnapshotResponse {
+  option (alpha_message) = true;
+
+  // Contains all attributes of the newly created group snapshot.
+  // This field is REQUIRED.
+  VolumeGroupSnapshot group_snapshot = 1;
+}
+
+message VolumeGroupSnapshot {
+  option (alpha_message) = true;
+
+  // The identifier for this group snapshot, generated by the plugin.
+  // This field MUST contain enough information to uniquely identify
+  // this specific snapshot vs all other group snapshots supported by
+  // this plugin.
+  // This field SHALL be used by the CO in subsequent calls to refer to
+  // this group snapshot.
+  // The SP is NOT responsible for global uniqueness of
+  // group_snapshot_id across multiple SPs.
+  // This field is REQUIRED.
+  string group_snapshot_id = 1;
+
+  // A list of snapshots created.
+  // This field is REQUIRED.
+  repeated Snapshot snapshots = 2;
+
+  // Timestamp when the volume group snapshot is taken.
+  // This field is REQUIRED.
+  .google.protobuf.Timestamp creation_time = 3;
+}
+```
+
+##### CreateVolumeGroupSnapshot Errors
+
+If the plugin is unable to complete the CreateVolumeGroupSnapshot call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Group snapshot already exists but is incompatible | 6 ALREADY_EXISTS | Indicates that a group snapshot corresponding to the specified group snapshot `name` already exists but is incompatible with the specified `volume_id`. | Caller MUST fix the arguments or use a different `name` before retrying. |
+| Not enough space to create group snapshot | 13 RESOURCE_EXHAUSTED | There is not enough space on the storage system to handle the create group snapshot request. | Caller SHOULD fail this request. Future calls to CreateVolumeGroupSnapshot MAY succeed if space is freed up. |
```

Comment:
Should we have a FAILED_PRECONDITION error if multiple volumes are specified that CANNOT be snapshot together? Some SP may have volumes spread across different storage backends, availability zones, etc preventing this. Some SP also have requirements that for write-order consistent snapshots, the volumes would have to be set up to have that capability ahead of time.

Reply:
Yes, that's reasonable.

Reply:
Added: https://github.com/container-storage-interface/spec/pull/519/commits/b3e48c6e38be7c1aeea5149f5b75b8e3dd23a6d9